### PR TITLE
Split miner rewards v2

### LINKED ISSUE #5: Split miner commits
- custodied wallet keys are only accessible to the respective city (mayor, treasurer, etc)
- if Stacking is active:
  - 70% of the miner commits are distributed to Stackers of the citycoin
  - 30% of the miner commits are distributed to the custodied wallet for the city ("THE BAG")
- if Stacking is not active:
  - 100% of the miner commits are distributed to the custodied wallet for the city ("THE BAG")

### PR DESCRIPTION
**Pull request by whoabuddy**
This PR supersedes #20 and adds similar changes and closes #5.

Test for `get-total-supply` fails due to only 70% being in contract versus expected 100%.

Need to look at how to test against when stacking is active or not. (100% vs 70%)

### CODE REVIEW

**Comment by asteriabtc** [comment]
thank you! great job :)))

**Comment by whoabuddy** [comment]
I was able to import the new constants into the clarinet tests and get them to pass, but there are still some outstanding items that need to be addressed:

- `set-tokens-mined` in contract splits 70% without checking if stacking is active
- `SPLIT_STACKER_PERCENTAGE` in tests is used to pass tests, but may not model the exact expected outcomes anymore
- `CITY_CUSTODIED_WALLET` and `SPLIT_CITY_PERCENTAGE` are defined in tests, but not tested, need to review this as well
- this change breaks the native tests, and logic there will need to be revised as well

At this stage I just wanted to see if I could get the changes and tests to run correctly, not 100% sure this is the way to go yet. (Maybe 70% sure? 😜)

**Comment by whoabuddy** [comment]
I think I've got the mechanism figured out now, will take a fresh look tomorrow for optimizations then may need some help on the tests.

**Comment by whoabuddy** [comment]
This PR is ready for review! It ended up touching a lot more than expected, changes outlined below:

- updates `block-commit` map to include `amount`, `amount-to-stackers`, and `amount-to-city`
- adds read-only function `get-block-commit-to-stackers` and test
- adds read-only function `get-block-commit-to-city` and test
- updates `set-tokens-mined` to accept new `block-commit` values
  - updates `tokens-per-cycle` to include only ustx sent to stackers
- updates `mine-tokens` to calculate split based on if any ustx are stacked in a cycle
- updates tests based on all changes above
- adds new function to add-on tests for setting and testing `city-wallet` var and balances

**Review by asteriabtc — APPROVED** [review]

File: contracts/citycoin.clar [hunk 1]
```diff
@@ -620,12 +644,40 @@
...
+        ;; check if stacking is active
+        (if stacked-something
+            ;; transfer with split if active
+            (begin 
+                (unwrap-panic (stx-transfer? (/ (* SPLIT_STACKER_PERCENTAGE amount-ustx) u100) tx-sender (as-contract tx-sender)))
+                (unwrap-panic (stx-transfer? (/ (* SPLIT_CITY_PERCENTAGE amount-ustx) u100) tx-sender (var-get city-wallet)))
+            )
+            ;; transfer to custodied wallet if not active
+            (unwrap-panic (stx-transfer? amount-ustx tx-sender (var-get city-wallet)))
+        )
```
Comment (LNow) [inline contracts/citycoin.clar:680]: I think here we can reuse variables calculated few lines above.
```suggestion
        (if stacked-something
            ;; transfer with split if active
            (begin 
                (unwrap-panic (stx-transfer? amount-ustx-to-stacker tx-sender (as-contract tx-sender)))
                (unwrap-panic (stx-transfer? amount-ustx-to-city tx-sender (var-get city-wallet)))
            )
            ;; transfer to custodied wallet if not active
            (unwrap-panic (stx-transfer? amount-ustx-to-city tx-sender (var-get city-wallet)))
        )
```

**Review by LNow — CHANGES_REQUESTED** [review]
Great job!
Please remove `tests/test-keychain.json` and check if my suggestion is correct or not.
Appart form that it looks correct.

File: contracts/citycoin.clar (hunk 1 above)
Comment (whoabuddy) [reply contracts/citycoin.clar]: Fixed in 29f9f6f. Also used `amount-ustx-to-city` for the 100% transfer to be consistent with using computed variables.

**Review by LNow — APPROVED** [review]

**Review by asteriabtc — APPROVED** [review]

**Pull request merged**